rust/src/delta.rs

```diff
@@ -707,6 +706,69 @@ impl DeltaTable {
             .collect::<Vec<String>>())
     }
 
+    /// Whether a path should be hidden for delta-related file operations, such as Vacuum.
+    /// Names of the form partitionCol=[value] are partition directories, and should be
+    /// deleted even if they'd normally be hidden. The _db_index directory contains (bloom filter)
+    /// indexes and these must be deleted when the data they are tied to is deleted.
+    fn is_hidden_directory(&self, path_name: &str) -> Result<bool, DeltaTableError> {
+        Ok(
+            (path_name.starts_with(&self.storage.join_path(&self.table_path, "."))
+                || path_name.starts_with(&self.storage.join_path(&self.table_path, "_")))
+                && !path_name
+                    .starts_with(&self.storage.join_path(&self.table_path, "_delta_index"))
+                && !path_name
+                    .starts_with(&self.storage.join_path(&self.table_path, "_change_data"))
+                && !self
+                    .state
+                    .current_metadata
+                    .as_ref()
+                    .ok_or(DeltaTableError::NoMetadata)?
+                    .partition_columns
+                    .iter()
+                    .any(|partition_column| {
+                        path_name.starts_with(
+                            &self.storage.join_path(&self.table_path, partition_column),
+                        )
+                    }),
+        )
+    }
+
+    /// Run the Vacuum command on the Delta Table: delete files no longer referenced by a Delta table and are older than the retention threshold.
+    /// We do not recommend that you set a retention interval shorter than 7 days, because old snapshots and uncommitted files can still be in use by concurrent readers or writers to the table. If vacuum cleans up active files, concurrent readers can fail or, worse, tables can be corrupted when vacuum deletes files that have not yet been committed.
+    pub async fn vacuum(
+        &mut self,
+        retention_hours: u64,
+        dry_run: bool,
+    ) -> Result<Vec<String>, DeltaTableError> {
+        let tombstones_path = self.get_stale_files(retention_hours)?;
+
+        let mut tombstones = vec![];
+        let mut all_files = self.storage.list_objs(&self.table_path).await?;
+        while let Some(obj_meta) = all_files.next().await {
+            let obj_meta = obj_meta?;
```

Comment:
I think it would be better to strip table path from `obj_meta.path` right after here and only work with object path like how it's done in the scala implementation, this way, we can avoid all the `self.storage.join_path` calls, which can get expensive if we are dealing with 100k+ objects.